sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowExec.scala

```diff
@@ -153,19 +153,24 @@ case class WindowExec(
   }
 
   /**
-   * Collection containing an entry for each window frame to process. Each entry contains a frames'
-   * WindowExpressions and factory function for the WindowFrameFunction.
+   * Collection containing an entry for each window frame to process. Each entry contains a frame's
+   * [[WindowExpression]]s and factory function for the WindowFrameFunction.
    */
-  private[this] lazy val windowFrameExpressionFactoryPairs = {
+  private[this] lazy val windowFrameExpressionFactoryPairs:
+    Seq[(mutable.Buffer[WindowExpression], InternalRow => WindowFunctionFrame)] =
+  {
     type FrameKey = (String, FrameType, Option[Int], Option[Int])
     type ExpressionBuffer = mutable.Buffer[Expression]
-    val framedFunctions = mutable.Map.empty[FrameKey, (ExpressionBuffer, ExpressionBuffer)]
+    type WindowExpressionBuffer = mutable.Buffer[WindowExpression]
+    val framedFunctions = mutable.Map.empty[FrameKey, (WindowExpressionBuffer, ExpressionBuffer)]
 
     // Add a function and its function to the map for a given frame.
-    def collect(tpe: String, fr: SpecifiedWindowFrame, e: Expression, fn: Expression): Unit = {
+    def collect(
+      tpe: String, fr: SpecifiedWindowFrame, e: WindowExpression, fn: Expression): Unit =
```

Comment:
I'll revert it to not "pollute" the PR. It was much easier to read the code when the internals were less generic (that does nothing at all).